File: Listas Circular e Duplamente encadeadas/soma.py
# criacao do objeto
class No:
    """Esta classe representa um nó/elemento de uma lista encadeada."""

    def __init__(self, valor, proximo=None, anterior=None):
        self.valor = valor
        self.prox = proximo
        self.ant = anterior


class Lista:
    """Esta classe contem funções para a manipulação de lista encadeada."""
    inicio = None

    def inserirFim(self, novo: No):
        if self.inicio == None:
            self.inicio = novo
        else:
            ultimo = self.inicio
            while ultimo.prox != None:
                ultimo = ultimo.prox

            ultimo.prox = novo
            novo.ant = ultimo

    def mostrar(self):
        atual = self.inicio
        while atual != None:
            print(atual.valor, end=' ')
            atual = atual.prox

    def somar(self, inicio1: No, ultimo2: No, valor):
        aux1: No = inicio1
        aux2: No = ultimo2
        i = 0
        while aux2.prox != None:
            aux2 = aux2.prox
        for i in range(valor):
            soma = aux1.valor + aux2.valor
            aux = No(soma)
            self.inserirFim(aux)
            if aux1.prox is not None:
                aux1 = aux1.prox

            if aux2.ant is not None:
                aux2 = aux2.ant
            i += 1


numero = int(input())
while numero != 0: #para quando e zero
    lista1 = Lista()
    lista2 = Lista()
    soma = Lista()
    print()
    x = input().split()
    x = [int(i) for i in x]

    for i in range(numero):
        novo = No(x[i])
        lista1.inserirFim(novo)
    x = input().split()
    x = [int(i) for i in x]
    for i in range(numero):
        novo = No(x[i])
        lista2.inserirFim(novo)

    soma.somar(lista1.inicio, lista2.inicio, numero)
    print()
    soma.mostrar()
    print()
    numero = int(input()) 


# inicializa uma lista vazia

File: Listas Circular e Duplamente encadeadas/soma.c
#include <stdio.h>
#include <stdlib.h>
struct elemento
{
    int valor;
    struct elemento *proximo;
    struct elemento *anterior;
};
typedef struct elemento *Pno;
typedef struct elemento No;
void inserirFim(Pno *, Pno *);
int vazia(Pno);
Pno criarElemento(int);
void Soma(Pno, Pno, int);

int main(void)
{
    int numero, elem;
    Pno lista1 = NULL, lista2 = NULL, novo;

    scanf("%d", &numero);
    while (numero >= 0 && numero <= 1000)
    {

        for (int i = 0; i < numero; i++)
        {
            scanf("%d", &elem);
            novo = criarElemento(elem);
            inserirFim(&lista1, &novo);
        }

        for (int i = 0; i < numero; i++)
        {
            scanf("%d", &elem);
            novo = criarElemento(elem);
            inserirFim(&lista2, &novo);
        }

        Soma(lista1, lista2, numero);

        lista1 = NULL;
        lista2 = NULL;

        scanf("%d", &numero);
    }
    return 0;
}
int vazia(Pno inicio)
{
    if (inicio == NULL)
        return 1;
    else
        return 0;
}
Pno criarElemento(int valor)
{
    Pno novo;
    novo = (Pno)malloc(sizeof(No));

    novo->valor = valor;
    novo->proximo = NULL;
    novo->anterior = NULL;
    return novo;
}
void inserirFim(Pno *inicio, Pno *novo)
{
    Pno ultimo;

    if (vazia(*inicio))
        *inicio = *novo;
    else
    {
        ultimo = *inicio;
        while (ultimo->proximo != NULL)
            ultimo = ultimo->proximo;
        ultimo->proximo = *novo;
        (*novo)->anterior = ultimo;
    }
}
void Soma(Pno inicio1, Pno ultimo2, int valor)
{
    Pno aux1 = inicio1;
    Pno aux2 = ultimo2;

    while (aux2->proximo != NULL)
    {
        aux2 = aux2->proximo;
    }
    printf("\n");
    for (int i = 0; i < valor; i++)
    {
        int soma = aux1->valor + aux2->valor;
        printf("%d ", soma);

        if (aux1->proximo != NULL)
        {
            aux1 = aux1->proximo;
        }

        if (aux2->anterior != NULL)
        {
            aux2 = aux2->anterior;
        }
    }
    printf("\n");
}